crates/sui-replay/src/batch_replay.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::replay::LocalExec;
+use crate::types::ReplayEngineError;
+use futures::future::join_all;
+use parking_lot::Mutex;
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicBool, AtomicU64};
+use std::sync::Arc;
+use sui_config::node::ExpensiveSafetyCheckConfig;
+use sui_types::base_types::TransactionDigest;
+use tokio::time::Instant;
+use tracing::{error, info};
+
+pub async fn batch_replay(
+    tx_digests: impl Iterator<Item = TransactionDigest>,
+    batch_size: u64,
+    rpc_url: String,
+    expensive_safety_check_config: ExpensiveSafetyCheckConfig,
+    use_authority: bool,
+    terminate_early: bool,
+) {
+    let provider = Arc::new(TransactionDigestProvider::new(tx_digests));
+    let mut tasks = vec![];
+    let cur_time = Instant::now();
+    for _ in 0..batch_size {
+        let provider = provider.clone();
+        let expensive_safety_check_config = expensive_safety_check_config.clone();
+        let rpc_url_ref = rpc_url.as_str();
+        tasks.push(async move {
+            run_task(
+                provider,
+                rpc_url_ref,
+                expensive_safety_check_config,
+                use_authority,
+                terminate_early,
+            )
+            .await
+        });
+    }
+    let all_failed_transactions: Vec<_> = join_all(tasks).await.into_iter().flatten().collect();
+    info!(
+        "Finished replaying {} transactions, took {:?}",
+        provider.get_total_count(),
+        cur_time.elapsed()
+    );
+    if all_failed_transactions.is_empty() {
+        info!("All transactions passed");
+    } else {
+        error!("Some transactions failed: {:?}", all_failed_transactions);
+    }
+}
+
+struct TransactionDigestProvider {
+    digests: Mutex<VecDeque<TransactionDigest>>,
+    total_count: usize,
+    // Used for early termination when a transaction failed.
+    should_terminate: AtomicBool,
```

Comment:
A `CancellationToken` is a little more idiomatic for this, right?

Reply:
Can you describe more? We need each task to be able to cancel all other tasks. AtomicBool is the best way I can think of.

Reply:
`CancellationToken` is part of `tokio_utils`: https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html

It's doing the same thing, but it has functions built-in that you are currently building on top of AtomicBool`.